### PR TITLE
Add PLUGIN_OIDC_TOKEN_ID support to AWS S3

### PR DESCRIPTION
**Pull request by Ompragash**
Please read [CONTRIBUTING.md](CONTRIBUTING.md) for details on our [code of conduct](CODE_OF_CONDUCT.md), and the process for submitting pull requests to us.

Fixes #

## Proposed Changes

-
-
-

### Checklist

- [ ] Read the [**CONTRIBUTING**](CONTRIBUTING.md) document.
- [ ] Read the [**CODE OF CONDUCT**](CODE_OF_CONDUCT.md) document.
- [ ] Add tests to cover changes.
- [ ] Ensure your code follows the code style of this project.
- [ ] Ensure CI and all other PR checks are green OR
    - [ ] Code compiles correctly.
    - [ ] Created tests which fail without the change (if possible).
    - [ ] All new and existing tests passed.
- [ ] Add your changes to `Unreleased` section of [CHANGELOG](CHANGELOG.md).
- [ ] Improve and update the [README](README.md) (if necessary).
- [ ] Ensure [documentation](./DOCS.md) is up-to-date. The same file will be updated in [plugin index](https://github.com/drone/drone-plugin-index/blob/master/content/meltwater/drone-cache/index.md) when your PR is accepted, so it will be available for end-users at http://plugins.drone.io.

### CODE REVIEW

File: storage/backend/s3/s3.go [hunk 1]
```diff
@@ -84,36 +88,34 @@ func New(l log.Logger, c Config, debug bool) (*Backend, error) {
...
+    errCh := make(chan error, 1)  // Buffer the channel to prevent blocking in case of error
+
+    go func() {
+        defer close(errCh)
+        out, err := b.client.GetObjectWithContext(ctx, in)
+        if err != nil {
+            errCh <- fmt.Errorf("get the object, %w", err)
+            return
+        }
+        defer out.Body.Close()  // Ensure the body is closed after reading
```
Comment (raghavharness) [inline storage/backend/s3/s3.go:105]: Any reason to remove `defer internal.CloseWithErrLogf(b.logger, out.Body, "response body, close defer")`  and use this ?

File: storage/backend/s3/s3.go [hunk 2]
```diff
@@ -195,3 +197,35 @@ func assumeRole(roleArn, roleSessionName string) *credentials.Credentials {
 
 	return credentials.NewCredentials(stsProvider)
 }
+
+func assumeRoleWithWebIdentity(roleArn, roleSessionName, webIdentityToken string) *credentials.Credentials {
```
Comment (raghavharness) [inline storage/backend/s3/s3.go:201]: this method should return (*credentials.Credentials, error)

File: storage/backend/s3/s3.go [hunk 3]
```diff
@@ -195,3 +197,35 @@ func assumeRole(roleArn, roleSessionName string) *credentials.Credentials {
...
+    // Create a new session
+	sess, err := session.NewSession()
+	if err != nil {
+		fmt.Printf("failed to create AWS session: %v", err)
+		return nil
```
Comment (raghavharness) [inline storage/backend/s3/s3.go:206]: we should be returning error

File: storage/backend/s3/s3.go [hunk 4]
```diff
@@ -195,3 +197,35 @@ func assumeRole(roleArn, roleSessionName string) *credentials.Credentials {
...
+		WebIdentityToken: aws.String(webIdentityToken),
+		DurationSeconds:  aws.Int64(duration),
+	}
+
+	// Call the AssumeRoleWithWebIdentity function
+	result, err := svc.AssumeRoleWithWebIdentity(input)
+	if err != nil {
+		fmt.Printf("failed to assume role with web identity: %v", err)
+		return nil
+	}
```
Comment (raghavharness) [inline storage/backend/s3/s3.go:226]: ditto

File: storage/backend/s3/s3.go [hunk 5]
```diff
@@ -44,7 +43,12 @@ func New(l log.Logger, c Config, debug bool) (*Backend, error) {
 	if c.Key != "" && c.Secret != "" { // nolint:gocritic
 		conf.Credentials = credentials.NewStaticCredentials(c.Key, c.Secret, "")
 	} else if c.AssumeRoleARN != "" {
-		conf.Credentials = assumeRole(c.AssumeRoleARN, c.AssumeRoleSessionName)
+		if c.OIDCTokenID != "" {
+            // Assume role with OIDC
+            conf.Credentials = assumeRoleWithWebIdentity(c.AssumeRoleARN, c.AssumeRoleSessionName, c.OIDCTokenID)
```
Comment (raghavharness) [inline storage/backend/s3/s3.go:48]: let's get the error here, if it's not null then print it and return that error

File: storage/backend/s3/s3.go (hunk 2 above)
Comment (Ompragash) [reply storage/backend/s3/s3.go]: pushed the initial commit with the logic to return error but then changed it to match assumeRole method. I'll create a new PR for assumeRole error logic.

File: storage/backend/s3/s3.go (hunk 1 above)
Comment (Ompragash) [reply storage/backend/s3/s3.go]: at first tested by removing the dependency on the internal package by using buffered channel and a direct `defer` to close the response body to keep it simple but yeah reintroduced "internal.CloseWithErrLogf" possibly keeping the focus only on OIDC feature.

**Review by raghavharness — APPROVED** [review]
LGTM

**Pull request merged**